presto-native-sidecar-plugin/src/main/java/com/facebook/presto/sidecar/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,243 @@
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.common.ErrorCode;
+import com.facebook.presto.spi.ConnectorId;
+import com.facebook.presto.spi.ErrorCodeSupplier;
+import com.facebook.presto.spi.NodeManager;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.TableHandle;
+import com.facebook.presto.spi.WarningCollector;
+import com.facebook.presto.spi.plan.PlanChecker;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanVisitor;
+import com.facebook.presto.spi.plan.SimplePlanFragment;
+import com.facebook.presto.spi.plan.TableScanNode;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.List;
+
+import static com.facebook.presto.sidecar.nativechecker.NativePlanCheckerErrorCode.NATIVEPLANCHECKER_CONNECTION_ERROR;
+import static com.facebook.presto.sidecar.nativechecker.NativePlanCheckerErrorCode.NATIVEPLANCHECKER_RESPONSE_MISSING_BODY;
+import static com.facebook.presto.sidecar.nativechecker.NativePlanCheckerErrorCode.NATIVEPLANCHECKER_UNKNOWN_CONVERSION_FAILURE;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Uses the native sidecar to check verify a plan can be run on a native worker.
+ */
+public final class NativePlanChecker
+        implements PlanChecker
+{
+    private static final Logger LOG = Logger.get(NativePlanChecker.class);
+    private static final MediaType JSON_CONTENT_TYPE = MediaType.parse("application/json; charset=utf-8");
+    private static final JsonCodec<PlanConversionResponse> PLAN_CONVERSION_RESPONSE_JSON_CODEC = JsonCodec.jsonCodec(PlanConversionResponse.class);
+    public static final String PLAN_CONVERSION_ENDPOINT = "/v1/velox/plan";
+
+    private final NodeManager nodeManager;
+    private final JsonCodec<SimplePlanFragment> planFragmentJsonCodec;
+    private final OkHttpClient httpClient;
+
+    public NativePlanChecker(NodeManager nodeManager, JsonCodec<SimplePlanFragment> planFragmentJsonCodec)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.planFragmentJsonCodec = requireNonNull(planFragmentJsonCodec, "planFragmentJsonCodec is null");
+        this.httpClient = new OkHttpClient.Builder().build();
+    }
+
+    @Override
+    public void validate(PlanNode planNode, WarningCollector warningCollector)
+    {
+        // NO-OP, only validating fragments
+    }
+
+    @Override
+    public void validateFragment(SimplePlanFragment planFragment, WarningCollector warningCollector)
+    {
+        if (!planFragment.getPartitioning().isCoordinatorOnly() && !isInternalSystemConnector(planFragment.getRoot())) {
```

Comment:
Some system queries do run on the worker see https://github.com/prestodb/presto/pull/21416. You should still validate those plans.

Reply:
@aditi-pandit do you know where the filtering happens where we decide whether to run a system query on the coordinator or the worker?

Reply:
@rschlussel : All the SystemTables have a distribution from the getDistribution() method that is used to build splits https://github.com/prestodb/presto/blob/master/presto-spi/src/main/java/com/facebook/presto/spi/SystemTable.java#L29C18-L29C33 that makes the distributed execution work.

But from a planning perspective, I think System tables are planned as regular tables with some special handling like https://github.com/prestodb/presto/pull/21725
https://github.com/prestodb/presto/pull/21285

Reply:
Thanks for point that out. I can look into this further, how about doing this as a followup?

Reply:
@aditi-pandit  would it be correct then to send over any system tables that have ALL_NODES partitioning and the others would be excluded?